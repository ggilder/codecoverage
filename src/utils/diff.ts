interface FileDiff {
  filename: string
  addedLines: number[]
  deletedLines: number[]
}

export function parseGitDiff(diffOutput: string): FileDiff[] {
  const fileDiffs: FileDiff[] = []
  const lines = diffOutput.split('\n')

  let currentFileDiff: FileDiff | undefined
  let currentAddedLines: number[] = []
  let currentDeletedLines: number[] = []
  let seenHeaderLine = false
  let deletionCurrentLineNumber = 0
  let additionCurrentLineNumber = 0

  for (const line of lines) {
    if (line.startsWith('diff --git')) {
      // New file diff starts
      if (currentFileDiff) {
        currentFileDiff.addedLines = currentAddedLines
        currentFileDiff.deletedLines = currentDeletedLines
        fileDiffs.push(currentFileDiff)
      }

      currentFileDiff = {
        filename: getFilenameFromDiffHeader(line),
        addedLines: [],
        deletedLines: []
      }
      currentAddedLines = []
      currentDeletedLines = []
      seenHeaderLine = false
    } else if (line.startsWith('@@')) {
      // Header line
      seenHeaderLine = true
      const lineInfo = getLineInfoFromHeaderLine(line)
      deletionCurrentLineNumber = lineInfo.deletionStartingLineNumber
      additionCurrentLineNumber = lineInfo.additionStartingLineNumber
    } else if (line.startsWith('+') && seenHeaderLine) {
      // Added line
      currentAddedLines.push(additionCurrentLineNumber)
      additionCurrentLineNumber++
    } else if (line.startsWith('-') && seenHeaderLine) {
      // Deleted line
      currentDeletedLines.push(deletionCurrentLineNumber)
      deletionCurrentLineNumber++
    } else if (seenHeaderLine) {
      // Context line
      deletionCurrentLineNumber++
      additionCurrentLineNumber++
    }
  }

  // Add the last file diff
  if (currentFileDiff) {
    currentFileDiff.addedLines = currentAddedLines
    currentFileDiff.deletedLines = currentDeletedLines
    fileDiffs.push(currentFileDiff)
  }

  return fileDiffs
}

function getFilenameFromDiffHeader(header: string): string {
  // Extract the filename from the diff header
  const startIndex = header.indexOf(' a/') + 3
  const endIndex = header.indexOf(' b/', startIndex)
  const filename = header.substring(startIndex, endIndex)
  return filename
}

function getLineInfoFromHeaderLine(line: string): {
  deletionStartingLineNumber: number
  additionStartingLineNumber: number
} {
  // Extract the starting line numbers for each side of the diff
  const matches = line.match(/-(\d+),?(\d+)? \+(\d+),?(\d+)? @@/)
  if (matches && matches.length === 5) {
    const deletionStartingLineNumber = parseInt(matches[1], 10)
    const additionStartingLineNumber = parseInt(matches[3], 10)
    return {deletionStartingLineNumber, additionStartingLineNumber}
  }
  return {deletionStartingLineNumber: 0, additionStartingLineNumber: 0}
}
