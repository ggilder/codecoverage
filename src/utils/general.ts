/** input: dist/index.test.ts --> output: index */
export function getFileNameFirstItemFromPath(path: string): string | undefined {
  const rawFileName = path?.split('/')?.pop()
  return rawFileName?.split('.')?.[0]
}

export function filterCoverageByFile(coverage: CoverageParsed): CoverageFile[] {
  return coverage.map(item => ({
    fileName: item.file,
    missingLineNumbers: item?.lines?.details
      .filter(line => line.hit === 0)
      .map(line => line.line)
  }))
}

export type CoverageParsed = {
  file: string
  title: string
  lines: {
    found: number
    hit: number
    details: {
      line: number
      hit: number
      name: string
    }[]
  }
}[]

export type CoverageFile = {
  fileName: string
  missingLineNumbers: number[]
}
