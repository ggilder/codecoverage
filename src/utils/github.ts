import * as core from '@actions/core'
import * as diff from './diff'
import * as github from '@actions/github'
import {
  CoverageFile,
  LineRange,
  coalesceLineNumbers,
  intersectLineRanges
} from './general'
import {Octokit} from 'octokit'

export class GithubUtil {
  private client: Octokit

  constructor(token: string) {
    if (!token) {
      throw new Error('GITHUB_TOKEN is missing')
    }
    this.client = new Octokit({auth: token})
  }

  getPullRequestRef(): string {
    const pullRequest = github.context.payload.pull_request
    return pullRequest
      ? pullRequest.head.ref
      : github.context.ref.replace('refs/heads/', '')
  }

  async getPullRequestDiff(): Promise<PullRequestFiles> {
    const pull_number = github.context.issue.number
    const response = await this.client.rest.pulls.get({
      ...github.context.repo,
      pull_number,
      mediaType: {
        format: 'diff'
      }
    })
    // @ts-expect-error With mediaType param, response.data is actually a string, but the response type doesn't reflect this
    const fileLines = diff.parseGitDiff(response.data)
    const prFiles: PullRequestFiles = {}
    for (const item of fileLines) {
      prFiles[item.filename] = coalesceLineNumbers(item.addedLines)
    }

    return prFiles
  }

  /**
   * https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/workflow-commands-for-github-actions#setting-a-warning-message
   */
  annotate(annotations: Annotations[]): void {
    for (const ann of annotations) {
      console.log(
        `::${ann.annotation_level} file=${ann.path},line=${ann.start_line},endLine=${ann.end_line}::${ann.message}`
      )
    }
  }

  buildAnnotations(
    coverageFiles: CoverageFile[],
    pullRequestFiles: PullRequestFiles
  ): Annotations[] {
    const annotations: Annotations[] = []
    for (const current of coverageFiles) {
      // Only annotate relevant files
      const prFileRanges = pullRequestFiles[current.fileName]
      if (prFileRanges) {
        const coverageRanges = coalesceLineNumbers(current.missingLineNumbers)
        const uncoveredRanges = intersectLineRanges(
          coverageRanges,
          prFileRanges
        )

        // Only annotate relevant line ranges
        for (const uRange of uncoveredRanges) {
          const message =
            uRange.end_line > uRange.start_line
              ? 'These lines are not covered by a test'
              : 'This line is not covered by a test'
          annotations.push({
            path: current.fileName,
            start_line: uRange.start_line,
            end_line: uRange.end_line,
            annotation_level: 'warning',
            message
          })
        }
      }
    }
    core.info(`Annotation count: ${annotations.length}`)
    return annotations
  }
}

type Annotations = {
  path: string
  start_line: number
  end_line: number
  start_column?: number
  end_column?: number
  annotation_level: string
  message: string
}

type PullRequestFiles = {
  [key: string]: LineRange[]
}
