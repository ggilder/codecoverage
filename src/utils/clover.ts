import * as clover from '@cvrg-report/clover-json'
import * as fs from 'fs'
import * as path from 'path'
import {CoverageParsed} from './general'

export async function parseClover(
  coveragePath: string,
  workspacePath: string
): Promise<CoverageParsed> {
  if (!coveragePath) {
    throw Error('No Clover XML path provided')
  }
  if (!workspacePath) {
    throw Error('No workspace path provided')
  }
  const fileRaw = fs.readFileSync(coveragePath, 'utf8')
  const parsed = await clover.parseContent(fileRaw)

  for (const entry of parsed) {
    entry.file = path.relative(workspacePath, entry.path || entry.file)
  }

  return parsed
}
