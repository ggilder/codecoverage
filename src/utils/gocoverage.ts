import * as fs from 'fs'
import * as gocov from 'golang-cover-parse'
import * as path from 'path'
import * as readline from 'readline'
import {CoverageParsed} from './general'

export async function parseGoCoverage(
  coveragePath: string,
  goModPath: string
): Promise<CoverageParsed> {
  if (!coveragePath) {
    throw Error('No Go coverage path provided')
  }

  if (!goModPath) {
    throw Error('No Go module path provided')
  }

  const goModule = await parseGoModFile(goModPath)

  const fileRaw = fs.readFileSync(coveragePath, 'utf8')
  return new Promise((resolve, reject) => {
    gocov.parseContent(fileRaw, (err, result) => {
      if (err === null) {
        filterModulePaths(result, goModule)
        resolve(result)
      }
      reject(err)
    })
  })
}

function filterModulePaths(entries, moduleName: string): void {
  for (const entry of entries) {
    entry.file = path.relative(moduleName, entry.file)
  }
}

async function parseGoModFile(filePath: string): Promise<string> {
  const fileStream = fs.createReadStream(filePath)
  const rl = readline.createInterface({
    input: fileStream,
    crlfDelay: Infinity
  })
  // Note: we use the crlfDelay option to recognize all instances of CR LF
  // ('\r\n') in input.txt as a single line break.

  for await (const line of rl) {
    if (line.startsWith('module ')) {
      return line.slice(7)
    }
  }

  /* istanbul ignore next */
  return ''
}
