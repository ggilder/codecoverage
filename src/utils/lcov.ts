import * as NodeUtil from 'util'
import * as fs from 'fs'
import parser from 'lcov-parse'
import {CoverageParsed} from './general'

export async function parseLCov(lcovPath: string): Promise<CoverageParsed> {
  if (!lcovPath) {
    throw Error('No LCov path provided')
  }
  const parserPromise = NodeUtil.promisify(parser)
  const fileRaw = fs.readFileSync(lcovPath, 'utf8')
  return parserPromise(fileRaw) as CoverageParsed
}
