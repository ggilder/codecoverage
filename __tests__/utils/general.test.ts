import {test} from '@jest/globals'
import {getFixturePath} from '../fixtures/util'
import {parseLCov} from '../../src/utils/lcov'
import {getFileNameFirstItemFromPath,filterCoverageByFile} from '../../src/utils/general'

test('getFileNameFirstItemFromPath', function () {
  const testCases = [
    {
      input: 'a/b/c.test.ts',
      output: 'c'
    },
    {
      input: 'd.ts',
      output: 'd'
    },
    {
      input: '',
      output: undefined
    },
    {
      input: 'a/b/c',
      output: undefined
    }
  ]
  testCases.forEach(test => {
    expect(getFileNameFirstItemFromPath(test.input))
  })
})

test('filterCoverageByFile', async function () {
  const path = getFixturePath('lcov.info')
  const parsedLcov = await parseLCov(path)
  const output = filterCoverageByFile(parsedLcov)
  expect(output).toMatchSnapshot()
})
