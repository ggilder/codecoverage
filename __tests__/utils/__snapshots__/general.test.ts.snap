// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filterCoverageByFile 1`] = `
Array [
  Object {
    "fileName": "src/utils/general.ts",
    "missingLineNumbers": Array [
      3,
      4,
    ],
  },
  Object {
    "fileName": "src/utils/github.ts",
    "missingLineNumbers": Array [
      12,
      18,
      19,
      28,
      29,
      33,
      34,
      35,
      36,
      37,
      39,
      40,
      44,
      56,
      57,
      64,
      65,
      67,
      68,
      69,
      70,
      82,
      83,
    ],
  },
  Object {
    "fileName": "src/utils/lcov.ts",
    "missingLineNumbers": Array [
      6,
      7,
      9,
      10,
      11,
    ],
  },
]
`;
